docs/GDPs/gdp-0002-regions.rst

```diff
@@ -0,0 +1,275 @@
+
+
+Usage and Impact
+----------------
+
+This is an optional feature, but will be the accepted approach to specialize computation at points in the horizontal iteration space.
+
+
+Backward Compatibility
+----------------------
+
+This GDP aims to be fully backward-compatible.
+
+
+Detailed Description
+--------------------
+
+As introduced above, we propose adding a new ``with horizontal()`` context that specializes the stencil on a region of the horizontal axes bounds using ``region`` objects, which pass information to GT4Py about the horizontal iteration space through the indexing operator, similar to numpy arrays.
+
+
+Axis Offsets
+++++++++++++
+
+Regions computation is specified using `Axis Offsets`, which are defined in GT4Py by subscripting the axes (``I``, ``J``, and ``K``).
+These can be indexed, which returns the specific indices within a stencil relative to the compute origin. For example: ``I[0]`` is the first compute point, ``I[1]`` the second, and finally ``I[-1]`` is the last point in the stencil compute domain along the ``I`` axis.
+
+Stencil computation in the horizontal axes behaves differently than in the vertical because statements execute over an index space that may extend beyond the limits defined in the stencil compute domain.
+Such ``extents`` cannot be represented by merely subscripting axes, since for example ``I[-1]`` referes to the last compute domain index along the ``I`` axis, not the point before the beginning of it.
+Axis Offsets therefore interally hold another offset, which can be set by adding or subtracting from the subscript.
+For example ``I[0] - 2`` is itself an Axis Offsets that refers to 2 points before the start of the compute domain.
+
+Axis Offsets may be assigned to variables in Python and/or used as externals in GT4Py in order to define ``region``.
+If the external variable is set to ``None``, then any regions using that external are ignored.
+
+Region Specification
+++++++++++++++++++++
+
+``region`` is a keyword in GT4Py that, when subscripted using slices of axis offsets, defines the restricted computation.
+These form the arguments to ``with horizontal()``.
+
+As an example, ``region[I[0], :]`` specifies a restriction to the first compute point on the ``I`` axis, and no restriction in the ``J`` axis.
+``I[0]`` is a single point, so when converted to a slice is still the single point.
+The ``J`` axis simply has ``:``, which is an unrestricted slice, which GT4Py interprets as an unrestricted axis (behaves normally).
+
+The previous example introduced a key element of regional computation: There must not only be a way of specifying axis offsets outside the compute domain, but slices that extend to infinity in each direction (or alternatively, unrestricted endpoints of axes).
+GTScript interprets an unrestricted start or stop element as extending to infinity (or, unrestricted).
```

Comment:
Why is exactly that? Can you provide a short explanation / use case for this requirement?

Reply:
Sure. Good idea.